### PR TITLE
feat(dart_frog_cli): allow dev command to run with a custom host

### PR DESCRIPTION
**Pull request by KunjKanani**
Issue #912 

## Status

**READY**

## Description

Feature for running the server on a custom host like 0.0.0.0. So, I can connect physical devices to the backend via the WIFI/LAN network. 

**Requirements**

- [ ] --host support in dart_frog_cli.

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart [hunk 1]
```diff
@@ -43,6 +43,12 @@ class DevServerDomain extends DomainBase {
       );
     }
 
+    final host = request.params?['host'];
+
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:47]: **Nit:** Remove white space to keep consistency with entire file.

```suggestion
```

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart [hunk 2]
```diff
@@ -216,4 +223,10 @@ class DevServerDomain extends DomainBase {
     }
     _devServerRunners.clear();
   }
+
+  bool _isValidHost(dynamic host) {
+    return host is String &&
+        RegExp(r'^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$')
+            .hasMatch(host);
+  }
 }
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:232]: **Nit:** 
Since the method doesn't need to be a member of the `DevServerDomain` we can consider it having it as a function instead.

```suggestion
}

bool _isValidHost(dynamic host) {
  return host is String &&
      RegExp(r'^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$')
          .hasMatch(host);
}
```

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart [hunk 3]
```diff
@@ -43,6 +43,12 @@ class DevServerDomain extends DomainBase {
       );
     }
 
+    final host = request.params?['host'];
+
+    if (!_isValidHost(host)) {
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:48]: This makes a host parameter required. As far as I am concerned this is a breaking change in the Dart Frog daemon because it will break all daemon users currently using the dev server domain (for example the VS Code extension).

We should either consider making it optional or bumping the daemon version? (cc: @renancaraujo)

File: packages/dart_frog_cli/lib/src/dev_server_runner/dev_server_runner.dart [hunk 4]
```diff
@@ -95,6 +98,9 @@ class DevServerRunner {
   /// [Logger] instance used to wrap stdout.
   final Logger logger;
 
+  /// Which host the server should start on.
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/dev_server_runner/dev_server_runner.dart:101]: **Nit:**

In the command option help we are using "Which host the server should bind to.", whereas here we are using "Which host the server should start on.". I think we should ensure they are consistent.

I prefer (personal preference) the former one, since the shelf package also uses "bind" as the name of the serve constructor.

File: packages/dart_frog_cli/pubspec.yaml [hunk 5]
```diff
@@ -8,7 +8,7 @@ documentation: https://dartfrog.vgv.dev/docs/overview
 topics: [server, backend, shelf, dart-frog, cli]
 
 screenshots:
-  - description: 'Dart Frog logo.'
+  - description: "Dart Frog logo."
```
Comment (alestiago) [inline packages/dart_frog_cli/pubspec.yaml:11]: We should revert this change:
```suggestion
  - description: 'Dart Frog logo.'
```

**Review by alestiago — COMMENTED** [review]
Thanks for contributing! Looks good so far 💙 .

I have yet to look at the tests, I'll wait until this conversation is resolved because the resolution might change the test files.

---

Also, I left some "Nits", from [The Standard of Code Review](https://google.github.io/eng-practices/review/reviewer/standard.html#:~:text=Reviewers%20should%20always%20feel%20free,they%20could%20choose%20to%20ignore.), it means:
> Reviewers should always feel free to leave comments expressing that something could be better, but if it’s not very important, prefix it with something like “Nit: “ to let the author know that it’s just a point of polish that they could choose to ignore.

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart (hunk 3 above)
Comment (erickzanardo) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:48]: @alestiago isn't the host always required? Even if the use don't inform one, we need to assume it is localhost?

If the daemon don't already have a default for this, I would say that we should follow the path of assuming localhost as default instead of introducing a breaking change.

Comment (KunjKanani) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:48]: @alestiago, I think @erickzanardo is correct. We can add localhost if the host is null. 

`final host = request.params?['host'] ?? 'localhost';`

Comment (alestiago) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:48]: > @alestiago isn't the host always required?

I believe, as of today, the _dev_server.start_ method does not require the host; see the [docs](https://dartfrog.vgv.dev/docs/advanced/daemon#method-start).

>  We can add localhost if the host is null.

That will make it a non-breaking change, I'm happy with both approaches.

Comment (KunjKanani) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:48]: @alestiago I suggest that let's make it a non-breaking change. You can create a new issue for Deamon to introduce breaking change.

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart (hunk 1 above)
Comment (KunjKanani) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart]: Check [Commit](https://github.com/VeryGoodOpenSource/dart_frog/pull/924/commits/2ba0ff5f665d5c761d846e497377c5184204c536)

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart (hunk 2 above)
Comment (KunjKanani) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart]: Check [Commit](https://github.com/VeryGoodOpenSource/dart_frog/pull/924/commits/2ba0ff5f665d5c761d846e497377c5184204c536)

File: packages/dart_frog_cli/lib/src/dev_server_runner/dev_server_runner.dart (hunk 4 above)
Comment (KunjKanani) [reply packages/dart_frog_cli/lib/src/dev_server_runner/dev_server_runner.dart]: Check [commit](https://github.com/VeryGoodOpenSource/dart_frog/pull/924/commits/76d30d627c4a31612cb7ed27d99e6fc6eefdd8e1)

File: packages/dart_frog_cli/pubspec.yaml (hunk 5 above)
Comment (KunjKanani) [reply packages/dart_frog_cli/pubspec.yaml]: check [commit](https://github.com/VeryGoodOpenSource/dart_frog/pull/924/commits/6917ad91d6cc7dd375def6d975f216246855b625)

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart [hunk 6]
```diff
@@ -83,6 +88,7 @@ class DevServerDomain extends DomainBase {
 
     final devServerRunner = _devServerRunnerBuilder(
       logger: logger,
+      host: '$host',
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:91]: Unnecessary interpolation

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart [hunk 7]
```diff
@@ -217,3 +223,9 @@ class DevServerDomain extends DomainBase {
     _devServerRunners.clear();
   }
 }
+
+bool _isValidHost(dynamic host) {
+  return host is String &&
+      RegExp(r'^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$')
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:229]: will "localhost" pass this test?

File: packages/dart_frog_cli/lib/src/dev_server_runner/dev_server_runner.dart [hunk 8]
```diff
@@ -325,7 +331,7 @@ class DevServerRunner {
     await _codegen();
     await serve();
 
-    final localhost = link(uri: Uri.parse('http://localhost:$port'));
+    final localhost = link(uri: Uri.parse('http://$host:$port'));
     progress.complete('Running on $localhost');
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/dev_server_runner/dev_server_runner.dart:336]: Is this strictically a cosmetic change? how shelf will know the passed host?

File: packages/dart_frog_cli/lib/src/commands/dev/dev.dart [hunk 9]
```diff
@@ -103,12 +109,14 @@ class DevCommand extends DartFrogCommand {
   Future<int> run() async {
     _ensureRuntimeCompatibility(cwd);
 
+    final host = (results['host'] as String?) ?? _defaultHost;
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/commands/dev/dev.dart:112]: You already defined `defaultsTo` to `_defaultHost`, therefore `results['host']` will never be null.

File: packages/dart_frog_cli/lib/src/commands/dev/dev.dart [hunk 10]
```diff
@@ -37,10 +37,16 @@ class DevCommand extends DartFrogCommand {
...
+      )
+      ..addOption(
+        'host',
+        defaultsTo: _defaultHost,
+        help: 'Which host the server should bind to.',
       );
   }
 
   static const _defaultDartVmServicePort = '8181';
+  static const _defaultHost = '127.0.0.1';
```
Comment (renancaraujo) [inline packages/dart_frog_cli/lib/src/commands/dev/dev.dart:49]: DEfault today is "localhost"

File: packages/dart_frog_cli/test/src/commands/dev_server_runner/dev_server_runner_test.dart [hunk 11]
```diff
@@ -180,11 +183,41 @@ void main() {
...
+          sigint: sigint,
+          runProcess: (_, __) async => processResult,
+        );
+
+        await expectLater(devServerRunner.start(), completes);
+
+        expect(devServerRunner.isWatching, isTrue);
+        expect(devServerRunner.isServerRunning, isTrue);
+        expect(devServerRunner.isCompleted, isFalse);
+      });
```
Comment (renancaraujo) [inline packages/dart_frog_cli/test/src/commands/dev_server_runner/dev_server_runner_test.dart:213]: Not enough assertion for the tested use case. See the immediately following test, where the custom port number is verified.

**Review by renancaraujo — COMMENTED** [review]
Is this intended to be a cosmetic change? I believe the devserver should know the custom address.

File: packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart (hunk 7 above)
Comment (KunjKanani) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:229]: No.

Comment (renancaraujo) [reply packages/dart_frog_cli/lib/src/daemon/domain/dev_server_domain.dart:229]: Yes, in [this line](https://github.com/VeryGoodOpenSource/dart_frog/pull/924/files#diff-aa1e9275509af92caf516f301386d49ca1ba051870bdee3c26091c44e07730b3R46) we are passing `localhost` to this function in case of nullability.

**Comment by alestiago** [comment]
Hi @KunjKanani I'll be closing this Pull Request, since we usually give a two week period of inactivity before closing. Feel free to re-open if you would like to continue working on it! Thank you 💙 🙌
